### PR TITLE
Refactors ucrFactor control and the dialogs that use it

### LINKED ISSUE #6936: Levels/Labels dialog freezes when there's too many levels
From https://github.com/africanmathsinitiative/R-Instat/issues/6900#issuecomment-962964087

> I have now found another bug in this dialogue and am delighted, because this is within a topic we should investigate more generally.
> This one is also present in Version 0.7.2, so has not been introduced by recent changes.
> Again, please see first if you can repeat it!
> a) Open the diamonds dataset from the library. It is the first example in the ggplot2 package.
> b) It has 10 variables, of which 3 are (ordered) factors and the other 7 are numeric.
> c) Make those 7 numeric variables into factors. The one called price is what causes the problem, so you can start there if you prefer.
> d) Now use right-click and the Levels/Labels dialogue. Once in the dialogue you can look at any of the factor variables. I often look at just one, and then right-click on others as I examine them.
> e) They have between 127 and 554 levels - except for price and that has 11602 different levels.
> f) Close, or press OK on the Levels/Labels dialogue after price.
> g) Then I am stuck. The program freezes.

### PR DESCRIPTION
**Pull request by Patowhiz**
Fixes #6936
1. Enhances and refactors `ucrFactor`, for instance adds a feature that makes it be able to handle multiple parameters.
2. Optimises `ucrFactor` code, making it to reload quickly. To test how significant this is, use the dataset explained in the issue and reload Levels/labels dialog twice, the old version takes almost 'forever' and even 'hangs'.
3. Adapts the dialogs that use control to take advantage of its new functionalities.

### CODE REVIEW

**Comment by Patowhiz** [comment]
@rdstern you can start testing performance and functionality of this using @dannyparsons suggested dataset in issue #6936.
I'll write a detailed description of what has been done in this PR and why many dialogs have been changed after I'm done with 2 more dialogs;
1. Procurement > Corruption Risk Index (CRI) > Define Corruption Risk Index (CRI)
2. Structured > Survival > Define...

To get a good understanding of the performance difference in regards to wide datasets. Compare this PR with any latest changes from the master or old versions. 
Thanks

**Comment by Patowhiz** [comment]
@africanmathsinitiative/developers this is now ready for review. 
You can test performance and functionalities added in this PR using @dannyparsons suggested dataset in issue #6936. 
This PR refactors and enhances `ucrFactor` control.
It also has changes to the dialogs and sub dialogs that use the control;

1. Prepare > Column: Factor > Levels/Labels...
2. Prepare > Column: Factor > Recode Levels...
3. Prepare > Column: Factor > Reference Level...
4. Prepare > Column: Factor > Unused Levels
5. Prepare > Data Frame > Filter Rows... - Define New Filter
6. Prepare > Data Frame > Filter Rows... - Filter From Factors
7. Climatic > Check Data > QC Rainfall... - Omit Months
8. Procurement > Corruption Risk Index (CRI) > Define Corruption Risk Index (CRI)
9. Structured > Survival > Define...

Please note, changes to the above dialogs and sub dialogs are strictly aimed at adapting them to use the improved controlled. Careful testing needed.
Some of these dialogs need refactoring as well.

To get a good understanding of the performance difference in regards to wide datasets from a user perspective, please compare this PR with any latest changes from the master or old versions using the suggested data set.

@rdstern @lloyddewit in regards to `ucrFactor` control, the next step after this PR is to implement copy pasting and possibly make the control use the custom grid control, especially because of Linux platforms. 
Thanks.

**Review by rdstern — APPROVED** [review]
Looks good to me.  I have checked some, but not all the options.  I also tried with a regular sequence in the diamonds data that was simply 1 to 53000, so with 53000 levels.  That also works ok.  I don't know whether we want to have an upper limit, above which we give a warning.  Would that be easy?  It will hardly ever be needed?

Good to have a check of the code - I assume that's @lloyddewit ?

**Comment by Patowhiz** [comment]
@rdstern on the issue of upper limit we can do that in a separate PR (the selector may need that too). Ideally, this control could be made to read from structures created by @ChrisMarsh82 then those structures could implement the limits, just like the way the data view window reads from those structures. 

@rdstern in regards to testing, I would strongly suggest another user to test all options in these dialogs. Just incase I missed something, I did my own testing but would be good to have an extra set of eyes. This PR should not introduce a new bug especially to critically dialogs like the filter.

**Comment by lloyddewit** [comment]
@Ivanluv please could you review? thanks

**Review by Ivanluv — DISMISSED** [review]

File: instat/ucrFactor.vb [hunk 1]
```diff
@@ -14,649 +14,872 @@
...
 
-    Public Sub AddEditableColumns(strColumns As String())
-        lstEditableColumns.AddRange(strColumns)
-        ApplyColumnSettings()
+    Private Sub _ucrLinkedReceiver_ControlValueChanged(ucrChangedControl As ucrCore) Handles _ucrLinkedReceiver.ControlValueChanged
+        'Calling this subroutine here is necessary.
+        'However when reloading a dialog, the receiver always clears it's contents first before
+        'restoring the previous selected variable.
+        'This forces this control to never retain what was previously selected despite it's inbuilt ability to do so
+        'Once the receiver is refactored this comment can be removed.
```
Comment (lloyddewit) [inline instat/ucrFactor.vb:146]:
```suggestion
        'todo Once the receiver is refactored this comment can be removed.

```

File: instat/ucrFactor.vb [hunk 2]
```diff
@@ -14,649 +14,872 @@
...
-        RefreshFactorData()
-    End Sub
-
-    Public Sub SetAsMultipleSelector()
-        bIsSelector = True
-        bIsMultipleSelector = True
-        bIsGridColumn = False
-        RefreshFactorData()
+    Private Sub _grdSheet_BeforeCut(sender As Object, e As BeforeRangeOperationEventArgs) Handles _grdSheet.BeforeCut
+        MsgBox("Cutting is currently disabled. This feature will be included in future versions." & Environment.NewLine & "Try copying and deleting from one column cells at a time.", MsgBoxStyle.Information, "Cannot cut from cells")
```
Comment (lloyddewit) [inline instat/ucrFactor.vb:151]:
```suggestion
        MsgBox("Cutting is currently disabled. This feature will be included in future versions." &
               Environment.NewLine & "Try copying and deleting from one column cells at a time.",
               MsgBoxStyle.Information, "Cannot cut from cells")
```

File: instat/ucrFactor.vb [hunk 3]
```diff
@@ -14,649 +14,872 @@
...
 
-    Public Sub SetIsGridColumn(strColumnName As String)
-        bIsSelector = False
-        bIsMultipleSelector = False
-        bIsGridColumn = True
-        strColumnToGet = strColumnName
-        RefreshFactorData()
+    Private Sub _grdSheet_BeforeCellKeyDown(sender As Object, e As BeforeCellKeyDownEventArgs) Handles _grdSheet.BeforeCellKeyDown
+        If e.KeyCode = unvell.ReoGrid.Interaction.KeyCode.Delete OrElse e.KeyCode = unvell.ReoGrid.Interaction.KeyCode.Back Then
+            MsgBox("Deleting cells is currently disabled. This feature will be included in future versions." & Environment.NewLine & "To remove a cell's value, replace the value with NA.", MsgBoxStyle.Information, "Cannot delete cells.")
```
Comment (lloyddewit) [inline instat/ucrFactor.vb:157]: add line breaks?

File: instat/ucrFactor.vb [hunk 4]
```diff
@@ -14,649 +14,872 @@
...
-        If shtCurrSheet IsNot Nothing Then
-            shtCurrSheet.SetSettings(unvell.ReoGrid.WorksheetSettings.Edit_Readonly, Not bEditable)
+        Dim bValid As Boolean = True
+        'do levels entry validation
+        If _grdSheet.ColumnHeaders(e.Cell.Column).Text = DefaultColumnNames.Level Then
+            If Not IsNumeric(e.NewData) Then
+                MsgBox("Levels must be numeric values", MsgBoxStyle.Information, "Invalid Value")
+                bValid = False
+            ElseIf e.NewData.Contains(".") Then
+                MsgBox("Levels must not be decimal .", MsgBoxStyle.Information, "Invalid Value")
```
Comment (lloyddewit) [inline instat/ucrFactor.vb:177]:
```suggestion
                MsgBox("Levels must not be decimal", MsgBoxStyle.Information, "Invalid Value")

```

File: instat/ucrFactor.vb [hunk 5]
```diff
@@ -14,649 +14,872 @@
...
-                        If ucrChkLevels IsNot Nothing Then
-                            ucrChkLevels.Enabled = False
-                            ucrChkLevels.Checked = True
-                        End If
-                    Else
-                        shtCurrSheet.HideColumns(iLevelsCol, 1)
-                    End If
-                End If
-            End If
+    'sets the state of the control, linked receiver and grid properties
```
Comment (lloyddewit) [inline instat/ucrFactor.vb:197]: tag with `<summary>`?

File: instat/ucrFactor.vb [hunk 6]
```diff
@@ -14,649 +14,872 @@
...
+                                           DefaultColumnNames.Label,
+                                           DefaultColumnNames.Freq,
+                                           DefaultColumnNames.SelectorColumn}.Contains(i))
+        _extraColNames = lstExtraColNames
+
+        'todo. do further a developer check??
+        'for instance  parameter column names must be in overall paremeter list
+        'else throw a develper error 
+        FillGridWithNewDataSheet()
+        'OnControlValueChanged()
```
Comment (lloyddewit) [inline instat/ucrFactor.vb:232]: delete?

File: instat/ucrFactor.vb [hunk 7]
```diff
@@ -14,649 +14,872 @@
...
-                Else
-                    bSelect = False
-                End If
-                shtCurrSheet(i, iSelectorColumnIndex) = bSelect
-            Next
+    Private Sub FillGridWithNewDataSheet()
+
+        'check for linked receiver and empty receiver contents
+        'also check column type of the receiver if it is a factor.
+        'the alternative way of getting the column type is by calling (left here for later reference)
```
Comment (lloyddewit) [inline instat/ucrFactor.vb:307]:
```suggestion
        'the alternative way of getting the column type is by calling (todo left here for later reference)
```

File: instat/ucrFactor.vb [hunk 8]
```diff
@@ -14,649 +14,872 @@
...
+        'clear the cached varible name too.
+        'An important step to do incase variable data type has been changed 
+        strCurrFactorVariableName = ""
+
+
+        'important
+        'ucrFilter control implementation forced this line addition
+        'for some reason when ucrFactor.Visible = True is called when 
+        '_grdSheet = Nothing, then the reogrid throws a visbility error.
+        'removal of this Visibility setting can removed once ucrFilter has been fully refactored
```
Comment (lloyddewit) [inline instat/ucrFactor.vb:362]:
```suggestion
        'todo removal of this Visibility setting can removed once ucrFilter has been fully refactored
```

File: instat/ucrFactor.vb [hunk 9]
```diff
@@ -14,649 +14,872 @@
...
+        grdFactorData.Visible = False
+    End Sub
+
+    ''' <summary>
+    ''' creates a new worksheet add adds it to the grid control
+    ''' fills the new worksheet with the factor variable metadata 
+    ''' applies the appropriate settings to the worksheet
```
Comment (lloyddewit) [inline instat/ucrFactor.vb:369]: add <para> tags (else it's displayed as single continuous sentence)

File: instat/ucrFactor.vb [hunk 10]
```diff
@@ -14,649 +14,872 @@
...
+                                                bIncludeNALevel As Boolean,
+                                                hiddenColumnNames As IEnumerable(Of String),
+                                                extraColumnNames As IEnumerable(Of String),
+                                                editableColumnNames As IEnumerable(Of String)) As Boolean
+
+
+        'get the factor metadata from R
+        Dim expDataFrame As SymbolicExpression
+        Dim clsConvertToCharacter As New RFunction
+        Dim clsRFunctionGetFactorData As New RFunction
```
Comment (lloyddewit) [inline instat/ucrFactor.vb:393]:
```suggestion
        Dim clsConvertToCharFunction As New RFunction
        Dim clsGetFactorDataFunction As New RFunction
```

**Review by lloyddewit — CHANGES_REQUESTED** [review]
@Patowhiz This looks great, I especially liked all the comments.
I made some optional suggestions, see what you think.
I reviewed everything except `ucrFactor.vb`. I_ still need to review the 2nd part of `ucrFactor.vb`. I still need to review `CreateAndFillNewDataSheet()` and the code that follows it. I'll try and complete this review tonight or tomorrow.

**Review by lloyddewit — CHANGES_REQUESTED** [review]
@Patowhiz I finished my review

File: instat/ucrFactor.vb (hunk 9 above)
Comment (Patowhiz) [reply instat/ucrFactor.vb:369]: Done

File: instat/ucrFactor.vb (hunk 6 above)
Comment (Patowhiz) [reply instat/ucrFactor.vb:232]: Done.

File: instat/ucrFactor.vb [hunk 11]
```diff
@@ -14,649 +14,872 @@
...
+
+    ''' <summary>
+    ''' holds column names to be hidden in the grid sheet after filling the metadata
+    ''' </summary>
+    Private _hiddenColNames As IEnumerable(Of String) = {}
+
+    ''' <summary>
+    ''' holds extra column names that will be added to the sheet after filling the metadata
+    ''' </summary>
+    Private _extraColNames As IEnumerable(Of String) = {}
```
[the remark this answers is not in the document]
Comment (Patowhiz) [reply instat/ucrFactor.vb]: @lloyddewit yes I did think of the `HashSet`. It directly complements the conceptual intent of these collections.
My problem was .Net `HashSet` do not preserve order (how I wish they had a java equivalent `LinkedHashSet `) . But after looking at how I designed the code, these collections will technically never have their elements changed (no `remove()` will ever be called directly on the collections), so no order issues will ever occur.

I have now refactored the code to use the `HashSet`.

**Comment by Patowhiz** [comment]
@lloyddewit I have added your very useful suggestions. 
This is now ready for your review.

**Comment by lloyddewit** [comment]
> @lloyddewit I have added your very useful suggestions. This is now ready for your review.

@Patowhiz Great. I'm happy with all the comments set to resolved. There's also a block of unresolved comments (starting at `dlgRecodeFactor.vb`), did you get a chance to look at these?
Thanks

File: instat/dlgRecodeFactor.vb [hunk 12]
```diff
@@ -92,15 +92,14 @@ Public Class dlgRecodeFactor
 
         ucrInputAddNa.SetParameter(New RParameter("na_level", 1))
 
-        ucrFactorGrid.SetReceiver(ucrReceiverFactor)
-        ucrFactorGrid.SetAsViewerOnly()
-        ucrFactorGrid.bIncludeCopyOfLevels = True
-        ucrFactorGrid.AddEditableColumns({"New Label"})
+        ucrFactorGrid.SetAsNormalGridColumn(ucrReceiverFactor,
+                                            extraColNames:={"New Label"},
```
[the remark this answers is not in the document]
Comment (Patowhiz) [reply instat/dlgRecodeFactor.vb:96]: @lloyddewit Ooh yes. This should be a constant string. let me implement that.

[the remark this answers is not in the document]
Comment (Patowhiz) [reply instat/dlgRecodeFactor.vb:96]: Done.

File: instat/mdlCoreControl.vb [hunk 13]
```diff
@@ -113,4 +113,27 @@ Module mdlCoreControl
...
+    ''' </summary>
+    ''' <param name="enumerable">list of items to be returned as R vector command</param>
+    ''' <param name="bOnlyIfMultipleElement">Default is False. 
+    ''' If set as true and enumerable has only 1 element 
+    ''' then the element will be returned without the c() command.
+    ''' <para>for instance, an enumerable that contains "1" will be returned as string "1"</para> </param>
+    ''' <returns>If <paramref name="enumerable"/> contains at least one element, then returns an 
+    '''          R vector string command, else returns "".</returns>
+    Public Function GetRVector(enumerable As IEnumerable(Of String), Optional bOnlyIfMultipleElement As Boolean = False) As String
+        If enumerable Is Nothing OrElse enumerable.Count < 1 Then
```
Comment (Patowhiz) [inline instat/mdlCoreControl.vb:130]:
```suggestion
        If enumerable Is Nothing OrElse enumerable.Count = 0 Then
```

**Comment by Patowhiz** [comment]
@lloyddewit I think I have now made all your suggestions. Thanks.

File: instat/ucrFilter.vb [hunk 14]
```diff
@@ -90,6 +91,11 @@ Public Class ucrFilter
         VariableTypeProperties()
     End Sub
 
+    'todo. currently, this subroutine is called even before load event of this control is called
+    'the visibility setting it's toggling affects how load events of controls it contains are called
+    'this may cause unintended 'load' bugs to some controls like reogrids
+    'when refactoring the sequence of events that call this,
+    'making sure that this is only called during this control's loading event should be prioritised
```
Comment (Patowhiz) [inline instat/ucrFilter.vb:98]:
```suggestion
    'making sure that this is only called during this control's loading event should be prioritised
    'See issue #7408 comments

```

File: instat/ucrFactor.vb [hunk 15]
```diff
@@ -14,649 +14,857 @@
...
+        'clear the cached varible name too.
+        'An important step to do incase variable data type has been changed 
+        strCurrFactorVariableName = ""
+
+
+        'important
+        'ucrFilter control implementation forced this line addition
+        'for some reason when ucrFactor.Visible = True is called when 
+        '_grdSheet = Nothing, then the reogrid throws a visbility error.
+        'todo removal of this Visibility setting can removed once ucrFilter has been fully refactored
```
Comment (Patowhiz) [inline instat/ucrFactor.vb:347]:
```suggestion
        'todo removal of this Visibility setting can removed once ucrFilter has been fully refactored
        'see issue #7408  comments
```

**Comment by lloyddewit** [comment]
> @lloyddewit I think I have now made all your suggestions. Thanks.

@Patowhiz There are 8 open comments above. For each comment, please could you resolve, or state that you rejected the comment?
thanks

File: instat/ucrFactor.vb (hunk 11 above)
[the remark this answers is not in the document]
Comment (Patowhiz) [reply instat/ucrFactor.vb]: Done

File: instat/ucrFactor.vb [hunk 16]
```diff
@@ -14,649 +14,872 @@
...
+    Private Sub AddSelectorButtonCells(grdSheet As unvell.ReoGrid.Worksheet,
+                                              iSelectorColIndex As Integer,
+                                              bSingleSelection As Boolean)
+
+        If bSingleSelection Then
+            Dim radioButtonGroup As New unvell.ReoGrid.CellTypes.RadioButtonGroup
+            For iRowIndex As Integer = 0 To grdSheet.Rows - 1
+                'create the cell first by setting its value
+                grdSheet(iRowIndex, iSelectorColIndex) = False
+                'then cet the body to be a radio button
```
[the remark this answers is not in the document]
Comment (Patowhiz) [reply instat/ucrFactor.vb]: Done

File: instat/ucrFactor.vb (hunk 10 above)
Comment (Patowhiz) [reply instat/ucrFactor.vb]: Done

File: instat/ucrFactor.vb (hunk 8 above)
Comment (Patowhiz) [reply instat/ucrFactor.vb]: Done.

File: instat/dlgLabelsLevels.vb [hunk 17]
```diff
@@ -57,8 +53,17 @@ Public Class dlgLabelsLevels
         ucrReceiverLabels.SetIncludedDataTypes({"factor"}, bStrict:=True)
         ucrReceiverLabels.strSelectorHeading = "Factors"
 
-        ucrSelectorForLabels.SetParameter(New RParameter("data_name", 0))
-        ucrSelectorForLabels.SetParameterIsString()
+        Dim dctParamAndColNames As New Dictionary(Of String, String)
+        dctParamAndColNames.Add("new_labels", ucrFactor.DefaultColumnNames.Label)
+        'todo. commented out temporariry
```
[the remark this answers is not in the document]
Comment (Patowhiz) [reply instat/dlgLabelsLevels.vb]: Done.

File: instat/dlgLabelsLevels.vb [hunk 18]
```diff
@@ -92,13 +99,27 @@ Public Class dlgLabelsLevels
     End Sub
 
     Private Sub SetRCodeforControls(bReset As Boolean)
-        SetRCode(Me, ucrBase.clsRsyntax.clsBaseFunction, bReset)
-        TestOKEnabled()
+
+        ucrSelectorForLabels.SetRCode(clsViewLabelsFunction, bReset)
+        ucrReceiverLabels.SetRCode(clsViewLabelsFunction, bReset)
+
+        'todo. commented out temporariry
```
[the remark this answers is not in the document]
Comment (Patowhiz) [reply instat/dlgLabelsLevels.vb]: Done.

File: instat/dlgRecodeFactor.vb [hunk 19]
```diff
@@ -243,20 +242,35 @@ Public Class dlgRecodeFactor
...
-        strCurrentLabels = ucrFactorGrid.GetColumnAsList(ucrFactorGrid.strLabelsName, False)
-        strNewLabels = ucrFactorGrid.GetColumnAsList("New Label", False)
+    Private Sub ucrFactorGrid_GridContentReFilledFromR() Handles ucrFactorGrid.GridContentReFilledFromR
+        'copy all cell values of the 'Label' column into the 'New Label' column everytime the grid is refilled from R
+        'GridContentReFilledFromR event is always followed by ControlValueChanged so no need to raise ControlValueChanged event
+        ucrFactorGrid.SetCellValues("New Label",
+                                    ucrFactorGrid.GetCellValues(ucrFactor.DefaultColumnNames.Label, False),
+                                    bRaisedControlValueChangedEvent:=False)
+    End Sub
+    Private Sub ucrFactorGrid_ControlValueChanged(ucrChangedControl As ucrCore) Handles ucrFactorGrid.ControlValueChanged
```
[the remark this answers is not in the document]
Comment (Patowhiz) [reply instat/dlgRecodeFactor.vb:254]: Done.

File: instat/dlgLabelsLevels.vb [hunk 20]
```diff
@@ -92,13 +99,27 @@ Public Class dlgLabelsLevels
...
+
+            If ucrChkIncludeLevelNumbers.Checked AndAlso
+                Not ucrFactorLabels.IsColumnComplete({ucrFactor.DefaultColumnNames.Level}) Then
+                ucrBase.OKEnabled(False)
+            Else
+                ucrBase.OKEnabled(True)
+            End If
         Else
             ucrBase.OKEnabled(False)
         End If
```
[the remark this answers is not in the document]
Comment (Patowhiz) [reply instat/dlgLabelsLevels.vb]: Done.

File: instat/ucrFactor.vb (hunk 7 above)
Comment (Patowhiz) [reply instat/ucrFactor.vb]: Done.

File: instat/ucrFactor.vb [hunk 21]
```diff
@@ -14,649 +14,872 @@
...
+    ''' For instance when a factor variable is changed, it's default metadata is refilled from R
+    ''' </summary>
+    Public Event GridContentReFilledFromR()
+
+    ''' <summary>
+    ''' the receiver linked to this control.
+    ''' will be used to display the factor metadata of the variable contained in it
+    ''' <para>The receiver should contain a factor variable</para>
+    ''' </summary>
+    Private WithEvents _ucrLinkedReceiver As ucrReceiverSingle
```
[the remark this answers is not in the document]
Comment (Patowhiz) [reply instat/ucrFactor.vb:49]: Yes  I found that weird too.

**Comment by Patowhiz** [comment]
@lloyddewit I think I have now resolved all the comments. Thanks.

**Review by lloyddewit — APPROVED** [review]

**Comment by lloyddewit** [comment]
@rdstern if you can test/approve again, then we can merge, thanks

**Review by rdstern — APPROVED** [review]
@Patowhiz seems fine now - and fast.  There's relatively little to show for all this work, but it was really important.  Well done.  Good to merge and still to keep checking.

**Pull request merged**